compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/builders/ExpressionHelpers.kt

```diff
@@ -20,26 +20,15 @@ import org.jetbrains.kotlin.utils.addToStdlib.assertedCast
 
 val IrBuilderWithScope.parent get() = scope.getLocalDeclarationParent()
 
-inline fun IrBuilderWithScope.irLet(
-    value: IrExpression,
-    origin: IrStatementOrigin? = null,
-    nameHint: String? = null,
-    body: (VariableDescriptor) -> IrExpression
-): IrExpression {
-    val irTemporary = scope.createTemporaryVariable(value, nameHint)
-    val irResult = body(irTemporary.descriptor)
-    val irBlock = IrBlockImpl(startOffset, endOffset, irResult.type, origin)
-    irBlock.statements.add(irTemporary)
-    irBlock.statements.add(irResult)
-    return irBlock
-}
-
 inline fun IrBuilderWithScope.irLetS(
     value: IrExpression,
     origin: IrStatementOrigin? = null,
     nameHint: String? = null,
     body: (IrValueSymbol) -> IrExpression
 ): IrExpression {
+    if (value is IrGetValue)
+        return body(value.symbol)
```

Comment:
Looks like this change might not be correct in general, if the `value` is an access to a mutable variable which is changed inside the `body`. (Credit to @SvyatoslavScherbina for noticing.) Maybe we could use this logic in a separate method where it'd explicitly say that mutating the subject inside the body is not allowed (not sure how to name it though)?

Reply:
Nice catch! Since this is only an optimization, I've just gotten rid of it for now.

It might be better to have a separate cleanup pass for the IR to get rid of unnecessary temporaries instead.